File: C++/euler-99-largest-exponential.cpp/euler-99-largest-exponential.cpp
#include <iostream> // Input/output stream for console input and output
#include <fstream>  // File stream for file input and output
#include <sstream>  // String stream for string manipulation
#include <cmath>    // Math functions, including logarithm

class FindLine {
public:
    // Constructor to initialize the filename
    FindLine(const std::string& filename) : filename(filename) {}

    // Main function to run the FindLine program
    void run() {
        // Check if the file can be opene
        if (!openFile()) {
            std::cout << "Error opening file: " << filename << std::endl;
            return;
        }

        int linenum = 0;            // Variable to keep track of the current line number being processed
        int highestLine = 0;        // Variable to store the line number with the highest total calculated so far
        double highestTotal = 0.0;  // Variable to store the highest total calculated from the logarithmic operation


        std::string line;
        // Read each line from the file
        while (std::getline(file, line)) {
            linenum++;
            int base, exponent;
            // Split the line into base and exponent
            if (!split(line, base, exponent)) {
                std::cout << "Error occurred, couldn't convert to integers: " << line << std::endl;
                return;
            }

            // Calculate the total using logarithm
            double total = calcTotal(base, exponent);

            // Compare the current line total to the highest line total
            if (total > highestTotal) {
                highestLine = linenum;
                highestTotal = total;
            }
        }

        // Print the result based on the line number
        printResult(highestLine);
    }

private:
    std::string filename;
    std::ifstream file;

    // Open the file and return true if successful
    bool openFile() {
        file.open(filename);
        return file.is_open();
    }

    // Calculate the total using logarithm
    double calcTotal(int base, int exponent) {
        return static_cast<double>(exponent) * std::log(static_cast<double>(base));
    }

    // Print the result based on the line number
    void printResult(int line) {
        if (line == 0) {
            std::cout << "No line has been found in the file." << std::endl;
        } else {
            std::cout << line << std::endl;
        }
    }

    // Split the line into base and exponent
    bool split(const std::string& line, int& base, int& exponent) {
        std::istringstream iss(line);
        char comma;

        // Read the base, comma, and exponent from the line
        if (!(iss >> base >> comma >> exponent) || comma != ',' || iss.peek() != EOF) {
            return false;
        }

        return true;
    }
};

int main(int argc, char* argv[]) {
    // Check if the correct number of command-line arguments is provided
    if (argc != 2) {
        std::cout << "Usage: ./euler-99-largest-exponential <filename>" << std::endl;
        return 0;
    }

    // Create an instance of FindLine and run the program
    FindLine finder(argv[1]);
    finder.run();

    return 0;
}

File: Go/euler-99-largest-exponential.go/euler-99-largest-exponential.go
// Package main is the entry point for the program.
package main

import (
	"bufio"   // Bufio provides buffering for input/output operations, which is used for efficiently reading lines from a file.
	"fmt"     // Fmt is used for formatted printing to the console, specifically used for error handling.
	"math"    // Math provides mathematical functions. In this program, it is used for logarithmic calculations.
	"os"      // Os is used to access the command line arguments and handle file operations, which for reading file input.
	"strconv" // Strconv is used to convert strings to integers, allowing the program to process numerical values from the input file.
	"strings" // Strings provides string manipulation functions. Here, it is used to split lines based on commas for parsing.
)

// main is the entry point of the program.
func main() {

	// Check if a command-line argument (filename) is provided
	if len(os.Args) != 2 {
		fmt.Println("Usage: go run euler-99-largest-exponential.go <filename>")
		return
	}

	// Get the filename from the command-line argument
	filename := os.Args[1]

	// Open the .txt file
	file, err := os.Open(filename)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close() // Ensure the file is closed at the end

	// Create a scanner to read the file line by line
	scanner := bufio.NewScanner(file)
	linenum := 0
	highestLine := 0
	highestTotal := 0.0

	// Iterate through each line in the file
	for scanner.Scan() {
		linenum++
		line := scanner.Text()
		base, exponent, err := split(line)
		if err != nil {
			fmt.Println("Error occurred, couldn't convert to integers:", err)
			return
		}

		// Calculate the total using logarithm
		total := calcTotal(base, exponent)

		// Compare the current line total to the highest line total
		if total > highestTotal {
			highestLine = linenum
			highestTotal = total
		}
	}
	printResult(highestLine)
}

// split splits the line by the comma, converts the values into int, and checks for errors.
func split(line string) (int, int, error) {
	// Split the input line by a comma to get individual values
	values := strings.Split(line, ",")

	// Check if the input line contains exactly two values: Base and exponent
	if len(values) != 2 {
		return 0, 0, fmt.Errorf("Invalid input: %s", line)
	}

	// Convert the values into integers
	base, error1 := strconv.Atoi(values[0])
	exponent, error2 := strconv.Atoi(values[1])

	// Check for errors during conversion
	if error1 != nil {
		return 0, 0, error1
	} else if error2 != nil {
		return 0, 0, error2
	} else {
		return base, exponent, nil
	}
}

// calcTotal calculates the total using logarithm.
func calcTotal(base, exponent int) float64 {
	return float64(exponent) * math.Log(float64(base)) // Using the power rule of logarithms.
}

// printResult prints the result based on the line number.
func printResult(line int) {
	if line == 0 {
		fmt.Println("No line has been found in the file.")
	} else {
		fmt.Println(line)
	}
}
